packages/core/core/src/services/document-service/utils/bidirectional-relations.ts

```diff
@@ -0,0 +1,174 @@
+/* eslint-disable no-continue */
+import { keyBy } from 'lodash/fp';
+import { async } from '@strapi/utils';
+import type { UID, Schema } from '@strapi/types';
+
+interface LoadContext {
+  oldVersions: { id: string; locale: string }[];
+  newVersions: { id: string; locale: string }[];
+}
+
+/**
+ * Loads all bidirectional relations that need to be synchronized when content entries change state
+ * (e.g., during publish/unpublish operations).
+ *
+ * In Strapi, bidirectional relations allow maintaining order from both sides of the relation.
+ * When an entry is published, the following occurs:
+ *
+ * 1. The old published entry is deleted
+ * 2. A new entry is created with all its relations
+ *
+ * This process affects relation ordering in the following way:
+ *
+ * Initial state (Entry A related to X, Y, Z):
+ * ```
+ *   Entry A (draft)     Entry A (published)
+ *      │                     │
+ *      ├──(1)→ X            ├──(1)→ X
+ *      ├──(2)→ Y            ├──(2)→ Y
+ *      └──(3)→ Z            └──(3)→ Z
+ *
+ *   X's perspective:         Y's perspective:         Z's perspective:
+ *      └──(2)→ Entry A         └──(1)→ Entry A         └──(3)→ Entry A
+ * ```
+ *
+ * After publishing Entry A (without relation order sync):
+ * ```
+ *   Entry A (draft)     Entry A (new published)
+ *      │                     │
+ *      ├──(1)→ X            ├──(1)→ X
+ *      ├──(2)→ Y            ├──(2)→ Y
+ *      └──(3)→ Z            └──(3)→ Z
+ *
+ *   X's perspective:         Y's perspective:         Z's perspective:
+ *      └──(3)→ Entry A         └──(3)→ Entry A         └──(3)→ Entry A
+ *                           (all relations appear last in order)
+ * ```
+ *
+ * This module preserves the original ordering from both perspectives by:
+ * 1. Capturing the relation order before the entry state changes
+ * 2. Restoring this order after the new relations are created
```

Comment:
Does this belong here or in the contributor docs?

Reply:
open for any option, I personally prefer having this collocated with the code as the only moment a user will read this is when modifying the code

Reply:
Why not just include the JSdoc documentation into the contributor doc so all code level documentation is automatically added in the contributor doc ? 

If the comment explains how the function works specificaly it should stay in the code. if it explains how the overall sync system works this should be in a .md file